### PR TITLE
Gitlab Event Collector

### PR DESCRIPTION
**Pull request by gal-forer**
## Contributing to Cortex XSOAR Content
Make sure to register your contribution by filling the [contribution registration form](https://forms.gle/XDfxU4E61ZwEESSMA)

**The Pull Request will be reviewed only after the contribution registration form is filled.**

## Status
- [ ] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
[link to the issue](https://jira-hq.paloaltonetworks.local/browse/XDR-54346)

## Description
A few sentences describing the overall goals of the pull request's commits.

## Screenshots
Paste here any images that will help the reviewer

## Minimum version of Cortex XSOAR
- [ ] 6.0.0
- [ ] 6.1.0
- [ ] 6.2.0
- [ ] 6.5.0
- [x] 6.8.0 

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [x] No

## Must have
- [x] Tests
- [x] Documentation

### CODE REVIEW

File: Packs/GitLab/ModelingRules/GitlabModelingrules.yml [hunk 1]
```diff
@@ -0,0 +1,18 @@
+name: Gitlab Mapping Rule
+id: _gitlab_mapping_to_dataset
+fromversion: 3.8.0
+tags: []
+rules: ''
+schema: |-
+  {
+    "okta_sso_raw": {
```
Comment (yaakovi) [inline Packs/GitLab/ModelingRules/GitlabModelingrules.yml:8]: update schema

File: Packs/GitLab/ModelingRules/GitlabModelingrules.xif [hunk 2]
```diff
@@ -0,0 +1,16 @@
+[MODEL: model = Audit, dataset = gitlab_gitlab_raw]
+|alter        XDM.Audit.original_event_id=id
```
Comment (yaakovi) [inline Packs/GitLab/ModelingRules/GitlabModelingrules.xif:2]: no filter?
```suggestion
alter        XDM.Audit.original_event_id=id
```

File: Packs/GitLab/Integrations/GitLabEventCollector/README.md [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+4. Click **Test** to validate the URLs, tokens, and connection.
+## Commands
+You can execute these commands in a playbook.
+#### test-module
+***
+Integration command for testing.
+
+#### fetch-events
+***
+Command that is activated by the engine to fetch event.
```
Comment (yaakovi) [inline Packs/GitLab/Integrations/GitLabEventCollector/README.md:31]:
```suggestion
```

File: Packs/GitLab/Integrations/GitLabEventCollector/README.md [hunk 4]
```diff
@@ -0,0 +1,35 @@
+GITLAB log events collector
```
Comment (yaakovi) [inline Packs/GitLab/Integrations/GitLabEventCollector/README.md:1]: improve description

File: Packs/GitLab/Integrations/GitLabEventCollector/GitLabEventCollector_description.md [hunk 5]
```diff
@@ -0,0 +1,16 @@
+Gitlab Events collector XSIAM
```
Comment (yaakovi) [inline Packs/GitLab/Integrations/GitLabEventCollector/GitLabEventCollector_description.md:1]:
```suggestion
Gitlab Events collector
```

File: Packs/GitLab/Integrations/GitLabEventCollector/GitLabEventCollector.yml [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+  - 1 day
+  - 3 day
+  - 5 day
+  - 1 week
+  - 1 month
+  - 1 year
+- display: Push events to XSIAM on get events
+  name: push_events
+  required: false
+  type: 8
```
Comment (yaakovi) [inline Packs/GitLab/Integrations/GitLabEventCollector/GitLabEventCollector.yml:53]: It should be a command argument, not an instance configuration.

File: Packs/GitLab/Integrations/GitLabEventCollector/GitLabEventCollector.yml [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+  defaultvalue: 1 day
+  options:
+  - 1 minute
+  - 1 hour
+  - 1 day
+  - 3 day
+  - 5 day
+  - 1 week
+  - 1 month
+  - 1 year
```
Comment (yaakovi) [inline Packs/GitLab/Integrations/GitLabEventCollector/GitLabEventCollector.yml:49]:
```suggestion
- display: First fetch from API time
  name: after
  type: 0
  required: true
  defaultvalue: 1 day
```

File: Packs/GitLab/Integrations/GitLabEventCollector/GitLabEventCollector_description.md [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+-
+ Collects the events log for events and audit events provided by Gitlab API
+
+
+---
+
+* **Server URL** - The API domain URL for Okta.
+* **API key** - The request API key.
+* **Event types** - The event ttpes to get as a comma delimitered string.
```
Comment (yaakovi) [inline Packs/GitLab/Integrations/GitLabEventCollector/GitLabEventCollector_description.md:10]:
```suggestion
* **Event types** - A comma-separated list of event types.
```

File: Packs/GitLab/Integrations/GitLabEventCollector/GitLabEventCollector_description.md [hunk 9]
```diff
@@ -0,0 +1,16 @@
+Gitlab Events collector XSIAM
+-
+ Collects the events log for events and audit events provided by Gitlab API
+
+
+---
+
+* **Server URL** - The API domain URL for Okta.
+* **API key** - The request API key.
```
Comment (yaakovi) [inline Packs/GitLab/Integrations/GitLabEventCollector/GitLabEventCollector_description.md:9]: add link to documentation on how to generate the API key

File: Packs/GitLab/Integrations/GitLabEventCollector/GitLabEventCollector.py [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+        if command == 'gitlab-get-events':
+            command_results = CommandResults(
+                readable_output=tableToMarkdown('gitlab Logs', events, headerTransform=pascalToSpace),
+                raw_response=events,
+            )
+            return_results(command_results)
+        elif command == 'fetch-events':
+            demisto.setLastRun(get_events.get_last_run(events))
+            demisto_params['push_events'] = True
+        if demisto_params.get('push_events'):
```
Comment (yaakovi) [inline Packs/GitLab/Integrations/GitLabEventCollector/GitLabEventCollector.py:119]: The argument name should be `should_push_events`.

File: Packs/GitLab/Integrations/GitLabEventCollector/GitLabEventCollector.py [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+
+def main() -> None:  # pragma: no cover
+    demisto_params = demisto.params()
+    url = urljoin(demisto_params['url'], '/api/v4/')
+    events_collection_management = {
+        'groups_ids': argToList(demisto_params.get('group_ids', '')),
+        'projects_ids': argToList(demisto_params.get('project_ids', '')),
+        'event_types': ['groups', 'projects']
+    }
+    user_ids = demisto_params.get('user_ids', '').split(',')
```
Comment (yaakovi) [inline Packs/GitLab/Integrations/GitLabEventCollector/GitLabEventCollector.py:57]: What are the `user_ids`?

File: Packs/GitLab/Integrations/GitLabEventCollector/GitLabEventCollector.py [hunk 12]
```diff
@@ -0,0 +1,129 @@
...
+    headers = {'PRIVATE-TOKEN': demisto_params.get('api_key', {}).get('credentials', {}).get('password')}
```
Comment (yaakovi) [inline Packs/GitLab/Integrations/GitLabEventCollector/GitLabEventCollector.py:58]:
```suggestion
    headers = {'PRIVATE-TOKEN': demisto_params.get('api_key', {}).get('password')}
```

File: Packs/GitLab/Integrations/GitLabEventCollector/GitLabEventCollector.py [hunk 13]
```diff
@@ -0,0 +1,129 @@
...
+    last_run = demisto.getLastRun()
+    if ('groups', 'projects', 'events') not in last_run:
+        last_run = dateparser.parse(demisto_params['after'].strip()).isoformat()
+        last_run = {
+            'groups': last_run,
+            'projects': last_run,
+            'events': last_run,
+        }
+    else:
+        last_run = last_run
```
Comment (yaakovi) [inline Packs/GitLab/Integrations/GitLabEventCollector/GitLabEventCollector.py:73]: redundant
```suggestion
```

File: Packs/GitLab/Integrations/GitLabEventCollector/GitLabEventCollector.py [hunk 14]
```diff
@@ -0,0 +1,129 @@
...
+                get_events.client.page = 1
+                get_events.client.event_type = event_type
+                events.extend(get_events.run())
+        get_events.client.event_type = 'events'
+        for obj_id in user_ids:
+            get_events.client.request.url = url + f'users/{obj_id}/events'
+            get_events.client.page = 1
+            events.extend(get_events.run())
+        if command == 'test-module':
+            return_results('ok')
```
Comment (yaakovi) [inline Packs/GitLab/Integrations/GitLabEventCollector/GitLabEventCollector.py:99]: the `test-module` doesn't test anything.

File: Packs/GitLab/Integrations/GitLabEventCollector/GitLabEventCollector.py [hunk 15]
```diff
@@ -0,0 +1,129 @@
...
+            return
+        for event in events:
+            if 'details' in event:
+                for action in ['add', 'change', 'remove']:
+                    if action in event['details']:
+                        event['details']['action'] = f'{action}_{event["details"][action]}'
+                        event['details']['action_type'] = action
+                        event["details"]['action_category'] = event['details'][action]
+                        break
```
Comment (yaakovi) [inline Packs/GitLab/Integrations/GitLabEventCollector/GitLabEventCollector.py:108]: Move to a function.

File: Packs/GitLab/Integrations/GitLabEventCollector/GitLabEventCollector.py [hunk 16]
```diff
@@ -0,0 +1,129 @@
...
+            )
+            return_results(command_results)
+        elif command == 'fetch-events':
+            demisto.setLastRun(get_events.get_last_run(events))
+            demisto_params['push_events'] = True
+        if demisto_params.get('push_events'):
+            send_events_to_xsiam(events, demisto_params.get('vendor', 'gitlab'),
+                                 demisto_params.get('product', 'gitlab'))
+        else:
+            return_error(f'Command not found: {command}')
```
Comment (yaakovi) [inline Packs/GitLab/Integrations/GitLabEventCollector/GitLabEventCollector.py:123]:
```suggestion
```

**Review by yaakovi — CHANGES_REQUESTED** [review]

File: Packs/GitLab/ModelingRules/GitlabModelingrules.xif (hunk 2 above)
Comment (gal-forer) [reply Packs/GitLab/ModelingRules/GitlabModelingrules.xif:2]: There wasn't one in the mapping rules I got, Do you want me to ask Dror he already wrote the xql ? @yaakovi

File: Packs/GitLab/Integrations/GitLabEventCollector/GitLabEventCollector.py (hunk 11 above)
Comment (gal-forer) [reply Packs/GitLab/Integrations/GitLabEventCollector/GitLabEventCollector.py:57]: David send me the wrong url to the docs so I implemented according to it it was suppose to be all the events, I'll fix it tomorrow @yaakovi

File: Packs/GitLab/Integrations/GitLabEventCollector/GitLabEventCollector.py (hunk 14 above)
Comment (gal-forer) [reply Packs/GitLab/Integrations/GitLabEventCollector/GitLabEventCollector.py:99]: It creates the events before so if it got there it means that it worked

**Comment by ShirleyDenkberg** [comment]
@yaakovi Doc review completed.

**Review by yaakovi — APPROVED** [review]

**Pull request merged**